### PR TITLE
Add MQTT Notification agent to v2

### PR DESCRIPTION
**Pull request by nalipaz**
This code is quite similar to what I did in #1054. The code in that PR works perfectly, however the changes here i cannot seem to get working as I get a server error whenever using the test in the notification agent. Any input would be appreciated.

### CODE REVIEW

File: plexpy/config.py [hunk 1]
```diff
@@ -297,6 +297,29 @@ def bool_int(value):
     'MONITOR_REMOTE_ACCESS': (int, 'Monitoring', 0),
     'MONITORING_INTERVAL': (int, 'Monitoring', 60),
     'MONITORING_USE_WEBSOCKET': (int, 'Monitoring', 0),
+    'MQTT_ENABLED': (int, 'MQTT', 0),
```
Comment (JonnyWong16) [inline plexpy/config.py]: `config.py` not used for notification agents in v2. No need to change this file.

File: plexpy/notifiers.py [hunk 2]
```diff
@@ -184,6 +185,11 @@ def available_notification_agents():
                        'name': 'osx',
                        'id': AGENT_IDS['osx']
                        })
+    if MQTT().validate():
```
Comment (JonnyWong16) [inline plexpy/notifiers.py]: Not need for MQTT.

Add MQTT at line 136 above (I have them in alphabetical order).

```
               {'label': 'MQTT',
                'name': 'mqtt',
                'id': AGENT_IDS['mqtt']
                },
```

File: plexpy/notifiers.py [hunk 3]
```diff
@@ -3167,6 +3175,170 @@ def return_config_options(self):
...
+            self.mqtt_client.on_connect = self.on_connect
+            self.mqtt_client.on_publish = self.on_publish
+        except:
+            logger.error(u"PlexPy Notifiers :: Cannot load MQTT Notifications agent.")
+            pass
+
+        logger.info(u"PlexPy Notifiers :: MQTT initialized.")
+
+
+    def validate(self):
```
Comment (JonnyWong16) [inline plexpy/notifiers.py]: Not needed.

File: plexpy/notifiers.py [hunk 4]
```diff
@@ -3167,6 +3175,170 @@ def return_config_options(self):
...
+        self.loop_flag = 1
+        self.counter = 0
+        self.success = False
+        self.qos = 1
+        self.retain = False
+        self.client_id = 'plexpy'
+        logger.info(u"PlexPy Notifiers :: MQTT init.")
+
+        try:
+            self.mqtt = __import__("paho.mqtt.client", globals(), locals(), ['client'], 0)
```
Comment (JonnyWong16) [inline plexpy/notifiers.py]: `import` at the top of the file instead.

```
import paho.mqtt.client as mqtt
```

File: plexpy/notifiers.py [hunk 5]
```diff
@@ -3167,6 +3175,170 @@ def return_config_options(self):
...
+            self.mqtt_client = self.mqtt.Client(None, True, None, config['protocol'])
```
Comment (JonnyWong16) [inline plexpy/notifiers.py]:
```
self.mqtt_client = mqtt.Client(protocol=self.config['protocol'])
```

File: plexpy/notifiers.py [hunk 6]
```diff
@@ -3167,6 +3175,170 @@ def return_config_options(self):
...
+            status = 'Connection refused - bad username or password'
+        elif rc == 5:
+            status = 'Connection refused - not authorised'
+        else:
+            status = 'Connection refused - unknown error'
+
+        if rc == 0:
+            logger.info(u"PlexPy Notifiers :: MQTT connection: %s.", status)
+            logger.info(u"PlexPy Notifiers :: MQTT notification: Publishing message.")
+            (rc, mid) = self.mqtt_client.publish(self.topic, json.dumps(self.data), self.qos, self.retain)
```
Comment (JonnyWong16) [inline plexpy/notifiers.py]:
```
self.config['topic']
```

File: plexpy/notifiers.py [hunk 7]
```diff
@@ -3167,6 +3175,170 @@ def return_config_options(self):
...
+        self.success = False
+        self.qos = 1
+        self.retain = False
+        self.client_id = 'plexpy'
+        logger.info(u"PlexPy Notifiers :: MQTT init.")
+
+        try:
+            self.mqtt = __import__("paho.mqtt.client", globals(), locals(), ['client'], 0)
+            self.mqtt_client = self.mqtt.Client(None, True, None, config['protocol'])
+            self.mqtt_client.username_pw_set(config['username'], config['password'])
```
Comment (JonnyWong16) [inline plexpy/notifiers.py]:
```
self.mqtt_client.username_pw_set(self.config['username'], self.config['password'])
```

**Comment by nalipaz** [comment]
@JonnyWong16 I have made the suggested changes, let me know if you see other issues. I have not yet retested with these changes.

**Comment by JonnyWong16** [comment]
I have to figure out how to setup a MQTT server to test it.

**Comment by nalipaz** [comment]
@JonnyWong16 use one of these publicly available ones:
https://github.com/mqtt/mqtt.github.io/wiki/public_brokers

**Comment by JonnyWong16** [comment]
Can we use `paho.mqtt.publish.single` instead? https://pypi.python.org/pypi/paho-mqtt/1.1#single

Then we don't need all this connecting, looping, disconnecting, etc. The downside is we won't receive a success/failed result.

I'm testing it with http://www.hivemq.com/demos/websocket-client/ and it seems to work.

**Pull request merged**